lib/apisession/session.ts

```diff
@@ -0,0 +1,103 @@
+import Map = require('../util/map');
+import conf = require('../config');
+
+export = Session;
+
+class Session {
+    // PRIVATE VARIABLES
+    private _blpsess: BAPI.Session;
+    private _logger: bunyan.Logger;
+    private _seconds: number; // Last checked activity timestamp
+    private _expired: boolean = false;
+    private _activeSubscriptions: Map<Subscription> = new Map<Subscription>();
+    private _receivedSubscriptions: Map<Subscription> = new Map<Subscription>();
+
+    // PROPERTIES
+    get seconds(): number {
+        return this._seconds;
+    }
+
+    get expired(): boolean {
+        return this._expired;
+    }
+
+    get activeSubscriptions(): Map<Subscription> {
+        return this._activeSubscriptions;
+    }
+
+    get receivedSubscriptions(): Map<Subscription> {
+        return this._receivedSubscriptions;
+    }
+
+    // PUBLIC VARIABLES
+    public inUse: number = 0;
+    public lastPollId: number = undefined;
+    public lastSuccessPollId: number = undefined;
+
+    // CONSTRUCTOR
+    constructor(blpsess: BAPI.Session) {
+        this._blpsess = blpsess;
+        this._logger = bunyan.createLogger(conf.get('loggerOptions'));
+        this._seconds = curSeconds();  // Record the creation time
+
+        // Tear down the current session when the underlying blpSession terminated unexpectedly
+        this._blpsess.once('SessionTerminated', (): void => {
+            if (!this._expired) {
+                this._logger.debug('blpSession terminated unexpectedly. Terminate session.');
+                this.clear();
+            }
+        });
+    }
+
+    // PUBLIC FUNCTIONS
+    expire(): boolean {
+        if (!this.isExpirable()) {
+            return false;
+        }
+
+        // unsubscribe all active subscriptions when session expires
+        if (this._activeSubscriptions.size) {
+            try {
+                this._blpsess.unsubscribe(this._activeSubscriptions.values());
+            } catch (err) {
+                this._logger.error(err, 'Error Unsubscribing');
+            }
+            this._logger.debug('Unsubscribed all active subscriptions.');
+        }
+        this.clear();
+        this._expired = true;
+        this._logger.debug('Session expires.');
+        return true;
+    }
+
+    isExpirable(): boolean {
+        return !this.inUse;
+    }
+
+    renew(): void {
+        this._seconds = curSeconds();  // Record the renewed time
+    }
+
+    // PRIVATE FUNCTIONS
+    private clear(): void {
+        if (this._receivedSubscriptions.size) {
+            var subscriptions = this._receivedSubscriptions.values();
+            subscriptions.forEach((sub: Subscription): void => {
+                sub.removeAllListeners();
+            });
+        }
+        this._activeSubscriptions.clear();
+        this._receivedSubscriptions.clear();
+        this.lastPollId = this.lastSuccessPollId = undefined;
+    }
+}
+
+function curSeconds(): number
```

Comment:
Let's move this before the class so we know where this function got introduced when reading the class.